src/Symfony/Component/Form/Extension/Core/Type/MultiStepType.php

```diff
@@ -0,0 +1,174 @@
+                }
+
+                return $value;
+            })
+            ->setDefault('current_step', static function (Options $options): string {
+                /** @var string $firstStep */
+                $firstStep = array_key_first($options['steps']);
+
+                return $firstStep;
+            })
+            ->setRequired('next_step')
+            ->setAllowedTypes('next_step', ['string', 'null'])
+            ->setDefault('next_step', function (Options $options): ?string {
+                return array_keys($options['steps'])[$this->currentStepIndex($options['current_step'], $options['steps']) + 1] ?? null;
+            })
+            ->setNormalizer('next_step', static function (Options $options, ?string $value): ?string {
+                if (null === $value) {
+                    return null;
+                }
+
+                if (!\array_key_exists($value, $options['steps'])) {
+                    throw new InvalidOptionsException(\sprintf('The next step "%s" does not exist.', $value));
+                }
+
+                return $value;
+            })
+            ->setRequired('previous_step')
+            ->setAllowedTypes('previous_step', ['string', 'null'])
+            ->setDefault('previous_step', function (Options $options): ?string {
+                return array_keys($options['steps'])[$this->currentStepIndex($options['current_step'], $options['steps']) - 1] ?? null;
+            })
+            ->setNormalizer('previous_step', static function (Options $options, ?string $value): ?string {
+                if (null === $value) {
+                    return null;
+                }
+
+                if (!\array_key_exists($value, $options['steps'])) {
+                    throw new InvalidOptionsException(\sprintf('The previous step "%s" does not exist.', $value));
+                }
+
+                return $value;
+            });
+
+        $resolver->setDefaults([
+            'hide_back_button_on_first_step' => false,
+            'button_back_options' => [
+                'label' => 'Back',
+            ],
+            'button_next_options' => [
+                'label' => 'Next',
+            ],
+            'button_submit_options' => [
+                'label' => 'Finish',
+            ],
+        ]);
+
+        $resolver->setAllowedTypes('hide_back_button_on_first_step', 'bool');
+        $resolver->setAllowedTypes('button_back_options', 'array');
+        $resolver->setAllowedTypes('button_submit_options', 'array');
+    }
+
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $currentStep = $options['steps'][$options['current_step']];
+
+        if (\is_callable($currentStep)) {
+            $currentStep($builder, $options);
+        } elseif (\is_string($currentStep)) {
+            $builder->add($options['current_step'], $currentStep);
+        }
+
+        $builder->add('back', SubmitType::class, [
```

Comment:
I'm not a fan of buttons within form type objects. I prefer their integration into the template.
There should be a way for the template to know when each button has to be displayed or not with the help of the `isFirstStep` or `isLastStep` methods.
Also, this will remove the need of the `hide_back_button_on_first_step`, `button_back_options`, `button_next_options` and `button_submit_options` options. WDYT?